api/api-users-v3/src/main/java/com/thoughtworks/go/apiv3/users/UsersControllerV3.java

```diff
@@ -0,0 +1,198 @@
+
+        if (user.equals(new NullUser())) {
+            throw new RecordNotFoundException();
+        }
+
+        UserToRepresent toRepresent = getUserToRepresent(user, roleConfigService.getRolesForUser(Collections.singletonList(user.getUsername())));
+        return writerForTopLevelObject(req, res, writer -> UserRepresenter.toJSON(writer, toRepresent));
+    }
+
+    public String create(Request req, Response res) throws Exception {
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        User user = buildUserEntityFromRequestBody(req, false);
+
+        if (!userService.findUserByName(user.getName()).equals(new NullUser())) {
+            throw haltBecauseEntityAlreadyExists(jsonWriter(user, new HashMap<>()), "User", user.getName());
+        }
+
+        return saveUserAndRenderResult(req, res, result, user, user.getName());
+    }
+
+    public String patchUser(Request req, Response res) throws Exception {
+        String username = req.params("login_name");
+
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        User userFromRequest = buildUserEntityFromRequestBody(req, true);
+
+        if (userService.findUserByName(username).equals(new NullUser())) {
+            throw haltBecauseNotFound();
+        }
+
+        if (isRenameAttempted(req, userFromRequest)) {
+            throw haltBecauseRenameOfEntityIsNotSupported("User");
+        }
+
+        return saveUserAndRenderResult(req, res, result, userFromRequest, username);
+    }
+
+    public String deleteUser(Request req, Response res) throws Exception {
+        String username = req.params("login_name");
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        userService.deleteUser(username, result);
+        return renderHTTPOperationResult(result, req, res);
+    }
+
+    public String bulkDelete(Request req, Response res) throws Exception {
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        JsonReader jsonReader = GsonTransformer.getInstance().jsonReaderFrom(req.body());
+        List<String> users = jsonReader.readStringArrayIfPresent("users").orElse(Collections.emptyList());
+
+        BulkDeletionFailureResult deletionFailureResult = userService.deleteUsers(users, result);
+
+        if (!deletionFailureResult.isEmpty()) {
+            res.status(result.httpCode());
+            return writerForTopLevelObject(req, res, outputWriter -> BulkDeletionFailureResultRepresenter.toJSON(outputWriter, deletionFailureResult, result));
+        }
+
+        return renderHTTPOperationResult(result, req, res);
+    }
+
+    private String saveUserAndRenderResult(Request req, Response res, HttpLocalizedOperationResult result, User user, String username) throws IOException {
+        userService.save(user, TriState.from(user.isEnabled()), TriState.from(user.isEmailMe()), user.getEmail(), user.getMatcher(), result);
```

Comment:
This is problematic. In the earlier API, if the boolean flag was not set, it implied not changing that property. The new implementation, it appears may possibly cause a `PATCH` api with no `enabled/email_me` option set to actually unset the value on the user record.

If the current implementation does handle that, then perhaps a test is required.